[MT-1748] Add support for custom variables and user identifier

sebastian-krajna opened the pull request:
Added support for sending user identifiers and custom variables to Contentsquare.

sebastian-krajna committed:
Update dependencies and add user identifier and custom variables support

- Updated CS_iOS_SDK to version 4.41.3 and tealium-swift to version 2.17.0.
- Introduced new functionalities for sending user identifiers and custom variables in Contentsquare.
- Added corresponding views and methods in the example app for user identifier and custom variables.
- Updated tests to cover new functionalities and ensure proper command processing.

sebastian-krajna committed:
Update project configuration and fix transaction ID key in tests

- Added CSCrashReporter.xcframework to the project.
- Removed the Embed Frameworks build phase and associated references.
- Updated the transaction ID key in ContentsquareInstanceTests to "transaction_id" for consistency.

sebastian-krajna committed:
Refactor custom variable handling and update example app

- Improved the validation logic for custom variables in ContentsquareInstance by introducing a new parseIndex method.
- Changed the example app to use a local contentsquare.json file instead of a remote URL.
- Updated the CustomVarsView to initialize with a default screen name and adjusted the indices of custom variables for consistency.
- Removed the IDEWorkspaceChecks.plist file as it is no longer needed.

sebastian-krajna committed:
Update version to 2.4.0 in podspec and ContentsquareConstants.swift

sebastian-krajna committed:
Enhance validation for custom variable index in ContentsquareInstance

- Added a guard clause to ensure the custom variable index is within the valid range of 1 to 20, improving error handling and robustness of the custom variable parsing logic.

Enricoza commented on Sources/ContentsquareInstance.swift:
```diff
@@ -51,6 +53,37 @@ public class ContentsquareInstance: ContentsquareCommand {
+            }
+            
+            guard let name = json["name"] as? String, !name.isEmpty,
+                  let value = json["value"] as? String, !value.isEmpty else {
```
It looks like empty name and value are allowed (and internally replaced with `cs-empty`)

```suggestion
            guard let name = json["name"] as? String,
                  let value = json["value"] as? String else {
```

Enricoza commented on Sources/ContentsquareInstance.swift:
```diff
@@ -70,6 +103,19 @@ public class ContentsquareInstance: ContentsquareCommand {
         Contentsquare.optOut()
     }
+    
+    private func parseIndex(_ value: Any?) -> UInt32? {
+        switch value {
+        case let i as UInt32:
+            return i
+        case let i as Int where i >= 0:
+            return UInt32(i)
```
This can crash in case i is greater than `UInt32.max`.
You can use a safer check like so:

```suggestion
    private func parseIndex(_ value: Any?) -> UInt32? {
        switch value {
        case let i as any BinaryInteger:
            return UInt32(exactly: i)
```

Enricoza commented on Sources/ContentsquareRemoteCommand.swift:
```diff
@@ -67,6 +67,14 @@ public class ContentsquareRemoteCommand: RemoteCommand {
+            case .sendUserIdentifier:
+                guard let userId = payload[ContentsquareConstants.UserIdentifier.userIdentifier] as? String else { return }
+                contentsquareInstance.sendUserIdentifier(userId: userId)
+            case .sendCustomVars:
+                guard let screenName = payload[ContentsquareConstants.ScreenView.screenName] as? String,
+                      let customVarsArray = payload[ContentsquareConstants.CustomVars.customVars] as? [[String: Any]],
+                      !customVarsArray.isEmpty else { return }
+                contentsquareInstance.sendCustomVars(screenName: screenName, customVars: customVarsArray)
```
We should not create a new command here. We should add the customVars as an optional parameter to the `sendScreenView` command.

Enricoza commented on Sources/ContentsquareInstance.swift:
```diff
@@ -70,6 +103,19 @@ public class ContentsquareInstance: ContentsquareCommand {
         Contentsquare.optOut()
     }
+    
+    private func parseIndex(_ value: Any?) -> UInt32? {
```
Since we created this method, we could use it also for `sendDynamicVar` like so:

```swift
    public func sendDynamicVar(dynamicVar: [String: Any]) {
        dynamicVar.forEach { key, value in
            if let value = value as? String {
                let dynamicVar = DynamicVar(key: key, value: value)
                Contentsquare.send(dynamicVar: dynamicVar)
            } else if let value = parseIndex(value) {
                let dynamicVar = DynamicVar(key: key, value: value)
                Contentsquare.send(dynamicVar: dynamicVar)
            } else {
                print("Incorrect format of value: \(value). Value should be String or UInt32.")
            }
        }
    }
```

Enricoza commented on TealiumContentsquare.xcodeproj/project.pbxproj:
```diff
@@ -66,7 +57,8 @@
 				15430459277C6820004FF14A /* TealiumCore.xcframework in Frameworks */,
 				1543045A277C6820004FF14A /* TealiumRemoteCommands.xcframework in Frameworks */,
-				15430458277C6820004FF14A /* SwiftProtobuf.xcframework in Frameworks */,
+				9E6990102DDB423A0005BB8B /* CSSwiftProtobuf.xcframework in Frameworks */,
```
We added `CSSwiftProtobuf` but didn't remove `SwiftProtobuf` from everywhere in the project. 
Some references are left, leading to a compile error in the project due to it being in the Frameworks list when you open the project.

Enricoza commented on Tests/MockContentsquareInstance.swift:
```diff
@@ -11,45 +11,57 @@ import XCTest
+    var lastCustomVarsInfo: (screenName: String, customVars: [[String: Any]])?
+    
+    // Tracking calls
+    var didStopTracking = false
+    var didResumeTracking = false
+    var didForgetMe = false
+    var didOptIn = false
+    var didOptOut = false
```
This is fine here, but in general is not a necessary change to make. 
The counter helps to test that the method is not called more than once.

sebastian-krajna replied:
Hmm, okay. I remember when I was working on Facebook SDK, you said if we provide change there it should also appear in rest of the the remote commands. Should I give up on this?

Enricoza replied:
no in general yes, this specific one is not needed though.

The rest of the test improvement like checking the actual values with the arrays and similar are great to have and you should do on all Remote Commands. Thanks!

Enricoza commented on Tests/ContentsquareInstanceTests.swift:
```diff
@@ -15,69 +15,311 @@ import XCTest
 class ContentsquareInstanceTests: XCTestCase {
 
-    let contentsquareInstance = MockContentsquareInstance()
+    var contentsquareInstance: MockContentsquareInstance!
```
This is not needed. The entire ContentsquareInstanceTest class is recreated at every test. Can be reverted.

Enricoza commented on Tests/ContentsquareInstanceTests.swift:
```diff
@@ -15,69 +15,311 @@ import XCTest
         contentsquareCommand = ContentsquareRemoteCommand(contentsquareInstance: contentsquareInstance)
     }
+    
+    override func tearDown() {
+        contentsquareInstance = nil
+        contentsquareCommand = nil
+    }
```
This is also not necessary. The remote command is deleted after every test.

Enricoza commented on TealiumContentsquareExample/TealiumContentsquareExample/ContentView.swift:
```diff
@@ -25,6 +25,14 @@ struct ContentView: View {
+                    Image(systemName: "person.crop.circle")
+                    Text("User ID")
+                }
+                CustomVarsView().tabItem {
+                    Image(systemName: "list.bullet.rectangle")
+                    Text("Custom Vars")
+                }
```
This also shouldn't be here. We can add an optional section in the Screens page

Enricoza commented on Sources/ContentsquareInstance.swift:
```diff
@@ -51,6 +53,37 @@ public class ContentsquareInstance: ContentsquareCommand {
+                return nil
+            }
+            
+            guard index > 0 && index <= 20 else {
+                print("CustomVar index out of range (1-20): \(index)")
+                return nil
+            }
```
Why is there this check? 
`> 0` is unnecessary as index is an UInt32 that can't be negative.
is `<= 20` a limitation of contentsquare? I can't find it anywhere

sebastian-krajna replied:
https://docs.contentsquare.com/en/android/track-custom-variables/

> It is possible to save up to 20 distinct custom variables per screenview. If more are received, only the first 20 custom variables, based on their index value, will be kept.
> The index value of custom variable is used to determine which custom variable to be kept. Only index value between 1 and 20 will be taken into account.

Enricoza replied:
Fair. I think we can just send it to them even if the index is greater than 20 (unless it crashes). It looks like they will ignore it for now, but if in future they increase it to 25 we won't block them for no reason.

Enricoza reviewed (changes requested):
Good start but there's some changes required. Have to finish the review still but I think there's enough to work on already before I finish the review.

sebastian-krajna committed:
Refactor screen view handling and remove custom variables support

- Updated the `sendScreenView` method in `ContentsquareInstance` to accept optional custom variables, enhancing flexibility in tracking screen views.
- Removed the `sendCustomVars` method and associated constants, streamlining the API.
- Adjusted the example app to reflect these changes, including the removal of the `CustomVarsView`.
- Updated tests to validate the new screen view functionality with custom variables.

sebastian-krajna committed:
Remove unused bordered extension from View in ButtonViewModifier.swift

sebastian-krajna committed:
Refactor custom variable handling in Contentsquare

- Updated `ContentsquareConstants` to use plural names for custom variable keys.
- Enhanced `ContentsquareRemoteCommand` to convert custom variables from JSON arrays to a structured format.
- Modified example app to reflect changes in custom variable structure.
- Updated tests to validate new custom variable handling and ensure compatibility with the updated format.

Enricoza commented on Sources/ContentsquareRemoteCommand.swift:
```diff
@@ -52,7 +52,13 @@ public class ContentsquareRemoteCommand: RemoteCommand {
+                // Convert from object of arrays to array of objects (JSON mapping format)
+                if let customVarsFromJSON = payload[ContentsquareConstants.CustomVars.customVars] as? [String: Any] {
+                    customVars = customVarsFromArrays(customVarsFromJSON)
+                }
```
I think we should try to get the vars as an array of objects first anyway (it might come like that from the Tag) but then try with this object of arrays if the first one fails, following the same approach we used for firebase

Enricoza commented on Sources/ContentsquareConstants.swift:
```diff
@@ -48,8 +48,8 @@ public enum ContentsquareConstants {
-        static let value = "value"
+        static let indexes = "indexes"
+        static let names = "names"
+        static let values = "values"
```
We should keep the singular here and use it everywhere (in the sendScreenView instance method and the customVarsArray command method) instead of the string literal version of these.

Enricoza commented on Sources/ContentsquareRemoteCommand.swift:
```diff
@@ -85,4 +91,36 @@ public class ContentsquareRemoteCommand: RemoteCommand {
+            if let values = customVarArrays[ContentsquareConstants.CustomVars.values], values.count > i {
+                result[i]["value"] = values[i]
+            }
+        }
```
The reason why we filtered the keys in `normalizeCustomVarArrays` is precisely to avoid repeating these lines for all keys. 
Once the keys are unified and put in singular, we can just write this:
```suggestion
        for i in 0 ..< count {
            for key in customVarArrays.keys {
                if let values = customVarArrays[key], values.count > i {
                    result[i][key] = values[i]
                }
            }
        }
```

Enricoza reviewed (commented):
Just a couple more updates needed

sebastian-krajna committed:
Refactor custom variable keys and handling in Contentsquare

- Updated `ContentsquareConstants` to use singular names for custom variable keys.
- Modified `ContentsquareRemoteCommand` to handle custom variables as arrays of objects.
- Adjusted example app JSON structure to reflect new custom variable key names.
- Updated tests to validate the new custom variable handling and ensure compatibility with the updated format.

Enricoza reviewed (approved):
Good job, looks good!

sebastian-krajna committed:
Update scripts and configurations for CocoaPods publishing and xcframework handling

- Added shebang to `publish.sh` for better script execution.
- Improved echo statements in `publish.sh` for clarity and consistency.
- Updated `xcframeworks.sh` to use a certificate variable instead of a team name for code signing.
- Enhanced file paths in `xcframeworks.sh` for better handling of xcframeworks.
- Added `tealium-xcframeworks` directory to `.gitignore` to prevent unnecessary files from being tracked.
